File: TDDMicroExercises/CSharp/TirePressureMonitoringSystem.Tests/MockAlarm.cs
using System;

namespace TDDMicroExercises.TirePressureMonitoringSystem.Tests
{
    public class MockAlarm : IAlarm
    {
        private bool _alarmOnReturnValue;

        private int _expectedCheckCallsCount = 0;
        private int _actualCheckCallsCount = 0;

        private int _expectedAlarmOnCallsCount = 0;
        private int _actualAlarmOnCallsCount = 0;

        public void SetExpectedCallToCheck()
        {
            _expectedCheckCallsCount = 1;
        }

        public void SetExpectedGetAlarmOn(bool returnValue)
        {
            _expectedAlarmOnCallsCount = 1;
            _alarmOnReturnValue = true;    
        }

        public void Verify()
        {
            if (_expectedAlarmOnCallsCount != _actualAlarmOnCallsCount)
            {
                throw new ArgumentOutOfRangeException("AlarmOn calls count", _actualAlarmOnCallsCount,
                                                      " expected: " + _expectedAlarmOnCallsCount);
            }

            if (_expectedCheckCallsCount != _actualCheckCallsCount)
            {
                throw new ArgumentOutOfRangeException("Check calls count", _actualCheckCallsCount,
                                                      "expected: " + _expectedCheckCallsCount);
            }
        }

        void IAlarm.Check()
        {
            _actualCheckCallsCount += 1;
        }

        bool IAlarm.AlarmOn
        {
            get
            {
                _actualAlarmOnCallsCount += 1;
                return _alarmOnReturnValue;
            }
        }
    }
}

File: TDDMicroExercises/Java/TelemetrySystem/src/test/java/tddmicroexercises/tirepressuremonitoringsystem/MockAlarm.java
package tddmicroexercises.tirepressuremonitoringsystem;

public class MockAlarm implements IAlarm
{
    private boolean alarmOn;

    private int expectedCheckCallsCount = 0;
    private int actualCheckCallsCount = 0;

    private int expectedAlarmOnCallsCount = 0;
    private int actualAlarmOnCallsCount = 0;

    public void resetExpectedCallsCount()
    {
        expectedCheckCallsCount = 1;
    }

    public void resetAlarmToOn()
    {
        expectedAlarmOnCallsCount = 1;
        alarmOn = true;    
    }

    public void verify()
    {
        if (expectedAlarmOnCallsCount != actualAlarmOnCallsCount)
        {
        	//TODO - find a more appropriate exception
            throw new IllegalArgumentException("AlarmOn calls count" + actualAlarmOnCallsCount + 
            		" expected: " + expectedAlarmOnCallsCount);
        }

        if (expectedCheckCallsCount != actualCheckCallsCount)
        {
            throw new IllegalArgumentException("Check calls count" +  actualCheckCallsCount +
                                                  "expected: " + expectedCheckCallsCount);
        }
    }

    public void check()
    {
        actualCheckCallsCount += 1;
    }

    public boolean isAlarmOn()
    {
            actualAlarmOnCallsCount += 1;
            return alarmOn;
    }
}

